AppController/djinn.rb

```diff
@@ -5427,6 +5440,111 @@ def scale_deployment
     }
   end
 
+  # Removes autoscaled nodes from the deployment as long as they are not running
+  # any AppServers and the minimum number of user specified machines are still
+  # running in the deployment.
+  def scale_down_instances
+    num_scaled_down = 0
+    # If we are already at the minimum number of machines that the user specified,
+    # then we do not have the capacity to scale down. 
+    max_scale_down_capacity = @nodes.length - Integer(@options['min_images'])
+    if max_scale_down_capacity == 0
+      Djinn.log_warn("We are already at the minimum number of user specified machines," +
+        "so will not be scaling down")
+      return
+    end
+
+    # Also, don't scale down if we just scaled up or down.
+    if Time.now.to_i - @last_scaling_time < (SCALEUP_THRESHOLD * 
+        SCALE_TIME_MULTIPLIER * DUTY_CYCLE)
+      Djinn.log_info("Not scaling down right now, as we recently scaled " +
+        "up or down.")
+      return
+    end
+
+    get_all_appengine_nodes.reverse_each { |node_ip|
+      # If we have already scaled down to our maximum capacity
+      # then return.
+      if num_scaled_down == max_scale_down_capacity
+        return
+      end 
+      
+      hosted_apps = []
+      @apps_loaded.each { |app_name|
+        @app_info_map[app_name]['appengine'].each { |location|
+          host, port = location.split(":")
+          if host == node_ip
+            hosted_apps << "#{app_name}:#{port}"
+          end
+        }
+      }
+      
+      unless hosted_apps.empty?
+        Djinn.log_debug("The node #{node_ip} has these AppServers " +
+        "running: #{hosted_apps}")
+        next
+      end
+
+      # Right now, only the autoscaled machines are started with just the 
+      # appengine role, so we check specifically for that during downscaling
+      # to make sure we only downscale the new machines added.
+      node_to_remove = nil
+      @state_change_lock.synchronize {
+        @nodes.each{ |node|
```

Comment:
Another check that we can add is the index number, since the min_images is set at the beginning (at least it should be set to be the number of nodes in the original layout in the AppScalefile) any nodes after @nodes[min_images + x] is been added afterward.